internal/database/inmemory.go

```diff
@@ -261,26 +248,88 @@ func CalculateAverage(samples []uplink.GPUStatSample) uplink.GPUStatSample {
 		MaxGraphicsClock:  sumMaxGraphicsClock / n,
 		MemoryClock:       sumMemoryClock / n,
 		MaxMemoryClock:    sumMaxMemoryClock / n,
-		Time:              minTime, // Use the earliest time as the timestamp for the averaged sample
+		Time:              minTime,
 		RunningProcesses:  aggregatedProcesses,
 	}
 
 	return averagedSample
 }
 
 func (m *inMemory) NewMachine(machine broadcast.NewMachine) error {
-	// TODO: add actual functionality. This was just to make the code compile
-	return errors.New("NOT IMPLEMENTED FOR IN-MEMORY DB")
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	if _, exists := m.lastSeen[machine.Hostname]; exists {
+		return fmt.Errorf("machine with hostname %s already exists", machine.Hostname)
+	}
+
+	m.lastSeen[machine.Hostname] = time.Now().Unix()
+
+	newMachine := broadcast.ModifyMachine{
+		Hostname: machine.Hostname,
+	}
+
+	if machine.Group != nil {
+		newMachine.Group = machine.Group
+	}
+
+	m.machines[machine.Hostname] = newMachine
+
+	return nil
 }
 
-func (m *inMemory) RemoveMachine(machine broadcast.RemoveMachine) (err error) {
-	// TODO: add actual functionality. This was just to make the code compile
-	return errors.New("NOT IMPLEMENTED FOR IN-MEMORY DB")
+func (m *inMemory) RemoveMachine(machine broadcast.RemoveMachine) error {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	var uuidToRemove string
+	for uuid, info := range m.infos {
+		if info.host == machine.Hostname {
+			uuidToRemove = uuid
+			break
+		}
+	}
+
+	if uuidToRemove == "" {
+		return fmt.Errorf("machine with hostname %s does not exist", machine.Hostname)
+	}
+
+	delete(m.lastSeen, machine.Hostname)
+	delete(m.infos, uuidToRemove)
+	delete(m.stats, uuidToRemove)
+
+	return nil
 }
 
 func (m *inMemory) UpdateMachine(changes broadcast.ModifyMachine) error {
-	// TODO: add actual functionality. This was just to make the code compile
-	return errors.New("NOT IMPLEMENTED FOR IN-MEMORY DB")
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	if _, exists := m.lastSeen[changes.Hostname]; !exists {
+		return fmt.Errorf("machine with hostname %s does not exist", changes.Hostname)
+	}
+
+	machine, exists := m.machines[changes.Hostname]
+	if !exists {
+		return fmt.Errorf("machine with hostname %s does not exist for update", changes.Hostname)
```

Comment:
ditto, I would just throw the same error as for the one above, and wrap it if need be.